### ISSUE TITLE
Missing Metrics plugin dependency makes "Add a new Cloud" crash

### ISSUE THREAD

**Issue by geoffroyjabouley** [jenkinsci/mesos-plugin#331]
Just after installing Mesos plugin on a fresh new Jenkins, i am not able to "Add a new Mesos Cloud".
It is generating a java exception because of missing dependency from Metrics plugin.

After installing Metrics plugin, i am able to configure a new Mesos Cloud.

It seems documentation is not accurate and Metrics plugin is not optional anymore. It this the case?

Jenkins: 2.107.3
Mesos plugin: 0.17

------------------------

```
Failed to load org.jenkinsci.plugins.mesos.MesosCloud
java.lang.ClassNotFoundException: com.codahale.metrics.Timer$Context
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1374)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1080)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
Caused: java.lang.NoClassDefFoundError: com/codahale/metrics/Timer$Context
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethods(Class.java:1975)
	at org.jvnet.hudson.annotation_indexer.Index$2$1.fetch(Index.java:103)
	at org.jvnet.hudson.annotation_indexer.Index$2$1.hasNext(Index.java:73)
	at org.jvnet.hudson.annotation_indexer.SubtypeIterator.fetch(SubtypeIterator.java:18)
	at org.jvnet.hudson.annotation_indexer.SubtypeIterator.hasNext(SubtypeIterator.java:28)
	at hudson.init.TaskMethodFinder.discoverTasks(TaskMethodFinder.java:56)
	at hudson.init.InitializerFinder.discoverTasks(InitializerFinder.java:33)
	at hudson.PluginManager$2$1$3.run(PluginManager.java:554)
	at org.jvnet.hudson.reactor.TaskGraphBuilder$TaskImpl.run(TaskGraphBuilder.java:169)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
	at jenkins.model.Jenkins$5.runTask(Jenkins.java:1062)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Sep 18, 2018 9:33:14 AM WARNING hudson.ExtensionFinder$Sezpoz scout
Failed to scout org.jenkinsci.plugins.mesos.MesosHealthCheckProvider
java.lang.ClassNotFoundException: jenkins.metrics.api.HealthCheckProvider
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1374)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1080)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
Caused: java.lang.NoClassDefFoundError: jenkins/metrics/api/HealthCheckProvider
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at jenkins.util.AntClassLoader.defineClassFromData(AntClassLoader.java:1140)
	at hudson.ClassicPluginStrategy$AntClassLoader2.defineClassFromData(ClassicPluginStrategy.java:847)
	at jenkins.util.AntClassLoader.getClassFromStream(AntClassLoader.java:1311)
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1364)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at jenkins.ClassLoaderReflectionToolkit.invoke(ClassLoaderReflectionToolkit.java:44)
	at jenkins.ClassLoaderReflectionToolkit._findClass(ClassLoaderReflectionToolkit.java:81)
	at hudson.PluginManager$UberClassLoader.findClass(PluginManager.java:1872)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at net.java.sezpoz.IndexItem.element(IndexItem.java:134)
Caused: java.lang.InstantiationException
	at net.java.sezpoz.IndexItem.element(IndexItem.java:146)
	at hudson.ExtensionFinder$Sezpoz.scout(ExtensionFinder.java:656)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:476)
	at hudson.ExtensionList.load(ExtensionList.java:366)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:304)
	at hudson.ExtensionList.getComponents(ExtensionList.java:169)
	at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1093)
	at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:83)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Sep 18, 2018 9:33:14 AM WARNING hudson.ExtensionFinder$Sezpoz _find
Failed to load org.jenkinsci.plugins.mesos.MesosHealthCheckProvider
java.lang.ClassNotFoundException: jenkins.metrics.api.HealthCheckProvider
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1374)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1080)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
Caused: java.lang.NoClassDefFoundError: jenkins/metrics/api/HealthCheckProvider
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at jenkins.util.AntClassLoader.defineClassFromData(AntClassLoader.java:1140)
	at hudson.ClassicPluginStrategy$AntClassLoader2.defineClassFromData(ClassicPluginStrategy.java:847)
	at jenkins.util.AntClassLoader.getClassFromStream(AntClassLoader.java:1311)
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1364)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1327)
	at sun.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at jenkins.ClassLoaderReflectionToolkit.invoke(ClassLoaderReflectionToolkit.java:44)
	at jenkins.ClassLoaderReflectionToolkit._findClass(ClassLoaderReflectionToolkit.java:81)
	at hudson.PluginManager$UberClassLoader.findClass(PluginManager.java:1872)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at net.java.sezpoz.IndexItem.element(IndexItem.java:134)
Caused: java.lang.InstantiationException
	at net.java.sezpoz.IndexItem.element(IndexItem.java:146)
	at hudson.ExtensionFinder$Sezpoz._find(ExtensionFinder.java:620)
	at hudson.ExtensionFinder$Sezpoz.find(ExtensionFinder.java:609)
	at hudson.ExtensionFinder$GuiceFinder.<init>(ExtensionFinder.java:257)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at net.java.sezpoz.IndexItem.instance(IndexItem.java:181)
	at hudson.ExtensionFinder$Sezpoz._find(ExtensionFinder.java:634)
	at hudson.ExtensionFinder$Sezpoz.find(ExtensionFinder.java:609)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:482)
	at hudson.ExtensionList.load(ExtensionList.java:366)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:304)
	at hudson.ExtensionList.getComponents(ExtensionList.java:169)
	at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1093)
	at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:83)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by maselvaraj** [comment]
Hi @geoffroyjabouley 
This commit has added metrics: https://github.com/jenkinsci/mesos-plugin/commit/0d0f10eb6a16d336bb4ca55ff12aa2f8b72a7024 

It is one of the commits from mesosphere enhancements. I agree documentation needs to be updated.

**Comment by geoffroyjabouley** [comment]
Thanks @maselvaraj 

Might be also relevant to remove "resolution:=optional" for Metrics dependency in the Mesos plugin Manifest:
`Plugin-Dependencies: credentials:1.22,metrics:3.1.2.6;resolution:=optional`

**Comment by jeschkies** [comment]
Fixed in #333.
